### PR TITLE
Gymnasium migration

### PR DESCRIPTION
**Pull request by Rohan138**
## Types of changes

- [X] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Migrating mbrl-lib from ```gym=0.17``` all the way to a combination of ```gym=0.26.2``` and ```gymnasium=0.26.3```.

## How Has This Been Tested (if it applies)

Will need thorough testing to ensure I didn't break anything, and also to make sure that performance is retained after refactoring the `step` API to use `terminated` and `truncated`.

## Checklist

- [X] The documentation is up-to-date with the changes I made.
- [X] I have read the [**CONTRIBUTING**](../../CONTRIBUTING.md) document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by Rohan138** [comment]
Update: Most files have been changed and fixed, tests with `tests/core`, `tests/pybullet`, `tests/mujoco`, `tests/dmcontrol`.

**Comment by Rohan138** [comment]
@luisenp @natolambert this should be ready for review.
- Note: I decided to add the `--follow-imports=skip` flag to mypy for now, typing the whole codebase correctly with all the gymnasium and gym wrappers around dmcontrol and pybullet envs is too large a refactor, hope that's okay?
- @raghavauppuluri13 is currently running learning experiments with this and the current `main` branch to verify that the migration does not affect performance; we'll put up results in a day or two.

File: mbrl/env/ant_truncated_obs.py
Comment (natolambert) [inline mbrl/env/ant_truncated_obs.py:32]: Do any of the changes like this change the dynamics? The `.sim` objects were always mysterious.

File: mbrl/env/ant_truncated_obs.py
Comment (natolambert) [inline mbrl/env/ant_truncated_obs.py:48]: Do we need to add logic for environments to count number of steps for `done`, or is that handled elsewhere?

File: mbrl/env/ant_truncated_obs.py
Comment (natolambert) [inline mbrl/env/ant_truncated_obs.py:70]: is `self.model.nv` the right shape? My understanding is that standard_normal and randn take different shape formats?

File: mbrl/env/humanoid_truncated_obs.py
Comment (natolambert) [inline mbrl/env/humanoid_truncated_obs.py:65]: same as above.

File: mbrl/third_party/pytorch_sac_pranz24/sac.py
Comment (natolambert) [inline mbrl/third_party/pytorch_sac_pranz24/sac.py:86]: Maybe add a comment here saying what we ignore?

File: tests/mujoco/test_util.py
Comment (natolambert) [inline tests/mujoco/test_util.py:62]: FYI still TODO here.

**Review by natolambert — COMMENTED** [review]
Did a quick review, interesting to me. Left some feedback, which maybe can be addressed in comments.

File: mbrl/env/ant_truncated_obs.py
Comment (Rohan138) [reply mbrl/env/ant_truncated_obs.py:32]: I think that's just the API change between the `mujoco_py` vs `mujoco` migration, actually (Most of the math remains the same)

File: mbrl/env/ant_truncated_obs.py
Comment (luisenp) [reply mbrl/env/ant_truncated_obs.py:48]: There used to be ` TimeLimit` wrapper to take care of that. I imagine something similar is still being done here.

Comment (Rohan138) [reply mbrl/env/ant_truncated_obs.py:48]: This is handled by the `gymnasium.wrappers.TimeLimit` wrapper we wrap all our environments with whenever they're instantiatied.

I just dug into this a little deeper though, noticed a minor bug:
https://github.com/facebookresearch/mbrl-lib/blob/main/mbrl/util/mujoco.py#L100
In this one place, `max_episode_steps=1000` is hardcoded here, whereas everywhere else, it's `max_episode_steps=cfg.overrides.get("trial_length", 1000)` e.g. https://github.com/facebookresearch/mbrl-lib/blob/main/mbrl/util/env.py#L94

I'll file this one as an issue separately, fix it in another branch since it's an existing bug not directly related to the Gymnasium migration.

Comment (luisenp) [reply mbrl/env/ant_truncated_obs.py:48]: Good catch, filing issue sounds good. Thanks!

File: mbrl/env/ant_truncated_obs.py
Comment (Rohan138) [reply mbrl/env/ant_truncated_obs.py:70]: Can you clarify? The numpy doc seems like they're identical: https://numpy.org/doc/stable/reference/random/generated/numpy.random.randn.html

File: tests/mujoco/test_util.py
Comment (Rohan138) [reply tests/mujoco/test_util.py:62]: This is a *very* weird bug: HalfCheetah is the only one that's flaky, and fails about 1/3 times I test it. I'm getting mildly suspicious of the underlying gymnasium env at this point, might open an issue there in a bit.

Comment (Rohan138) [reply tests/mujoco/test_util.py:62]: Update: Testing with `for run in {1..10}; do pytest -rP tests/mujoco/test_util.py -k "test_freeze"; done`, of the 10 gym environments, only `Ant-v4`, `HalfCheetah-v4`, and `HumanoidStandup-v4` are failing.

Comment (luisenp) [reply tests/mujoco/test_util.py]: What error are you getting?

Comment (Rohan138) [reply tests/mujoco/test_util.py]:
```
env_name = 'gym___Ant-v4'

    def _freeze_mujoco_gym_env(env_name):
        handler = create_handler_from_str(env_name)
        env = handler.make_env_from_str(env_name)
        env.reset(seed=0)
    
        seen_obses = []
        seen_rewards = []
        actions = []
        num_steps = 100
    
        with handler.freeze(env):
            for _ in range(num_steps):
                action = env.action_space.sample()
                next_obs, reward, terminated, truncated, _ = env.step(action)
                seen_obses.append(next_obs)
                seen_rewards.append(reward)
                actions.append(action)
                if terminated or truncated:
                    break
    
        for a in actions:
            next_obs, reward, _, _, _ = env.step(a)
            ref_obs = seen_obses.pop(0)
            ref_reward = seen_rewards.pop(0)
            np.testing.assert_array_almost_equal(next_obs, ref_obs)
>           assert reward == pytest.approx(ref_reward)
E           assert 0.14005772492059787 == 0.14005799237443473 ± 1.4e-07
E             comparison failed
E             Obtained: 0.14005772492059787
E             Expected: 0.14005799237443473 ± 1.4e-07

tests/mujoco/test_util.py:37: AssertionError
```

Comment (Rohan138) [reply tests/mujoco/test_util.py]: The values are always slightly off, but I've tried decreasing the tolerance on the comparison and it only becomes less flaky, not 100% passing. Plus I really shouldn't need to do that logically, based on what the `freeze` is doing.

File: mbrl/env/ant_truncated_obs.py
Comment (natolambert) [reply mbrl/env/ant_truncated_obs.py:70]: Okay, that's clearer than what I found. You're right.

**Comment by raghavauppuluri13** [comment]
Benchmarking the PR against main. Here are the reward curves for some selected. The planet algorithms are still training, but you can see they are generally similar.
image

**Comment by natolambert** [comment]
@raghavauppuluri13 lmk if you're compute limited, shouldn't be too hard for me to run some too.

**Comment by raghavauppuluri13** [comment]
> @raghavauppuluri13 lmk if you're compute limited, shouldn't be too hard for me to run some too.

Do we want to benchmark all the env/algo combos in the examples? If so, that would be great. I've got a helper script

```
#!/bin/bash

main_envs=("pets_reacher" "pets_pusher" "planet_cartpole_swingup" "planet_finger_spin" "mbpo_halfcheetah" "mbpo_inv_pendulum")
diff_envs=("pets_reacher" "pets_pusher" "planet_cartpole_swingup" "planet_finger_spin" "mbpo_half_cheetah_v4" "mbpo_inv_pendulum_v4")
algos=("pets" "pets" "planet" "planet" "mbpo" "mbpo")
device=("cuda:0" "cuda:0" "cuda:0" "cuda:0" "cuda:0" "cuda:0")
envs=$main_envs
tsp -S 6

for i in ${!envs[@]}; do
    if [[ "${algos[$i]}" == "planet" ]]; then
        tsp python -m mbrl.examples.main algorithm=${algos[$i]} dynamics_model=${algos[$i]} overrides=${envs[$i]} device=${device[$i]}
    else
        tsp python -m mbrl.examples.main algorithm=${algos[$i]} overrides=${envs[$i]} device=${device[$i]}
    fi
done
```

File: tests/mujoco/test_util.py
Comment (raghavauppuluri13) [reply tests/mujoco/test_util.py]: I made a simple reproducible example of this issue and it seems to be isolated to gymnasium. Opened an issue: https://github.com/Farama-Foundation/Gymnasium/issues/352

File: mbrl/env/humanoid_truncated_obs.py
Comment (raghavauppuluri13) [reply mbrl/env/humanoid_truncated_obs.py:65]: ^ also resolved with above

File: tests/mujoco/test_util.py
Comment (raghavauppuluri13) [reply tests/mujoco/test_util.py]: Given that this bug is gymnasium/mujoco specific, can we keep the flakey envs commented out and open a separate issue?

**Comment by raghavauppuluri13** [comment]
@natolambert @luisenp Any other validation steps before the pr is ready to merge? (I edited the benchmarks with the finished training curves)

File: tests/mujoco/test_util.py
Comment (luisenp) [reply tests/mujoco/test_util.py]: Does this issue break the CEM-based controller (on gt dynamics)? Or is it just affecting the test?

**Review by luisenp — CHANGES_REQUESTED** [review]
Let's also bump the version number and update the CHANGELOG. After that (pending my other question) we should be good to merge.

File: tests/mujoco/test_util.py
Comment (Rohan138) [reply tests/mujoco/test_util.py]: I don't think the freeze wrapper is used there (or anywhere, actually), so the controller should be fine.

Comment (luisenp) [reply tests/mujoco/test_util.py]: It's used by the controller for ground truth dynamics, under `diagnostics/control_env.py`. 

In [this](https://github.com/facebookresearch/mbrl-lib/blob/main/mbrl/diagnostics/control_env.py#L59) line we call `EnvHandler.rollout_env`, which in turn calls each class' `freeze` method [here](https://github.com/facebookresearch/mbrl-lib/blob/3c6553a43e8fea37b0695dacab91e3fefb147ad7/mbrl/util/env.py#L261). And for Mujoco this is defined [here](https://github.com/facebookresearch/mbrl-lib/blob/3c6553a43e8fea37b0695dacab91e3fefb147ad7/mbrl/util/mujoco.py#L71), which I believe is the same freeze called in the unit test.

Comment (raghavauppuluri13) [reply tests/mujoco/test_util.py]: > Does this issue break the CEM-based controller (on gt dynamics)? Or is it just affecting the test?
@luisenp 

From the tests, it seem that it affects the specific envs `Ant-v4, HalfCheetah-v4',HumanoidStandup-v4` as it doesn't affect the `Hopper-v4,Humanoid-v4`. So I can plan to run the `control_env.py` on all those envs and compare the results to `main` and see what happens. 

If it breaks on the known affected envs, I can open a separate issue after we merge this PR detailing that those envs do not work correctly and link the gymnasium issue and wait for that fix to resolve this issue.

Comment (luisenp) [reply tests/mujoco/test_util.py]: I agree with that plan. Thanks @raghavauppuluri13!

Comment (raghavauppuluri13) [reply tests/mujoco/test_util.py]: @luisenp @Rohan138  These are the results. One thing to note is that we're comparing the v2 versions with the v4 versions between main and diff respectively. This concludes that all are training, but the performance is slightly different between the versions. Does that look fine? Still worth making an issue to track the gymnasium issue that was made with these curves if people find slight issues with reproduction.
image

Comment (luisenp) [reply tests/mujoco/test_util.py]: Yeah, this is good. Thanks a lot @raghavauppuluri13! I want to give a final pass before merging, but I don't expect much changes to be necessary. I'll try to get this in by Thursday. Thanks!

File: mbrl/util/replay_buffer.py
Comment (luisenp) [inline mbrl/util/replay_buffer.py:548]: Are we sure about this? For this book keeping, it seems to me that we need the old meaning of `done`, that is, `terminated or truncated`. Otherwise, the info for a truncated trajectory will not be saved in the buffer, or be closed either, and the buffer will try to append transitions from the next trajectory to the truncated one. This doesn't affect sampling transitions, but it does affect the code used to sample full trajectories.

**Review by luisenp — CHANGES_REQUESTED** [review]
Left some final comments. We can merge after addressing these. Thanks!

**Comment by natolambert** [comment]
Awesome work all, have been following along happily.

File: mbrl/util/replay_buffer.py
Comment (Rohan138) [reply mbrl/util/replay_buffer.py:548]: Correct, fixing this

Comment (luisenp) [reply mbrl/util/replay_buffer.py:548]: The issue is that the trajectory is ["closed"](https://github.com/facebookresearch/mbrl-lib/blob/3c6553a43e8fea37b0695dacab91e3fefb147ad7/mbrl/util/replay_buffer.py#L502) only when `terminated = True`. The closing method is the only part of the code that saves the start/end indices to the list of trajectories that can be sampled, and the only method that [resets](https://github.com/facebookresearch/mbrl-lib/blob/3c6553a43e8fea37b0695dacab91e3fefb147ad7/mbrl/util/replay_buffer.py#L515) the start of the last trajectory, which makes it so that the next new trajectory is tracked correctly.

Comment (Rohan138) [reply mbrl/util/replay_buffer.py:548]: Ah, that makes sense. Fixed now.

**Comment by Rohan138** [comment]
@luisenp Could I get another review? Should be good to merge now, let me know if there's anything else I can change

**Pull request merged**
